### PR TITLE
Feature enumerables

### PR DESCRIPTION
**Pull request by moinkhanif**
We included our own methods in the extended Enumerable module:

*  #my_each
*  #my_each_with_index
*  #my_select
*  #my_all?
*  #my_any?
*  #my_none?
*  #my_count
*  #my_map - able to take in proc and/or a block
*  #my_inject
*  #multiply_els to test #my_inject

### CODE REVIEW

**Comment by kobiyoyo** [comment]
## Hi @moinkhanif       
#### Great Job so far, but there are a few more things required. 
### General
- [ ] `my_each` 
-  returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_each` returns `#<Enumerator>`
- [ ] `my_each_with_index`  
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_each_with_index` returns `#<Enumerator>`
- [ ] `my_select`  
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_select` returns `#<Enumerator>`
- [ ] `my_map` 
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_map` returns `#<Enumerator>`
- [ ] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`












- [ ] `puts [3,3,3,3,3].my_all?(2)`  should return false .
- [ ] Enumerable `my_any?` when no block or argument is given returns true if at least one of the collection is not false or nil.
- `p [3,3,3,3].my_any?(4)`should return false 
- `p [].my_any? ` should return false
-  `p %w[ant bear cat].my_any? { |word| word.length >= 4 }` should return true
- [ ] `my_none?`:
- When no block or argument is given returns true only if none of the collection members is true
`p  [nil, false, true, []].my_none?` - should return false
- When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`p ['hello'].my_none?(5)` - should return true

### Make changes and submit another code review request.
*Happy coding*

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @moinkhanif       
#### Great Job so far, but there are a few more things required. 
### General
- [x] `my_each` 
-  returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_each` returns `#<Enumerator>`
- [x] `my_each_with_index`  
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_each_with_index` returns `#<Enumerator>`
- [x] `my_select`  
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_select` returns `#<Enumerator>`
- [x] `my_map` 
- returns an Enumerator if no block is given:
- puts `[1, 2, 3].my_map` returns `#<Enumerator>`
- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`
- [x] `my_all`  
- when no block or argument is given returns true when none of the collection members are false or nil:
- `p [1, false, 'hi', []].my_all?` should return `false`.
- when a class is passed as an argument returns true if all of the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return `true`
-  when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['hello','flex'].my_all?(/d/)` should return `false`
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`p [3,3,3,3].my_all?(3)` should return `true`
- [x] `my_none?`:
- When no block or argument is given returns true only if none of the collection members is true
`p  [nil, false, true, []].my_none?`  should return `false`
- When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`p ['hello'].my_none?(5)` - should return `true`
- when a class is passed as an argument returns true if none of the collection is a member of such class
`p [3,3,3,3].my_none?(String)`  should return `true`
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
`p ['hello','milk','milo'].my_none?(/z/)` should return `true`

- [x] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(Integer)` should return `true`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('cat')` should return `true`



### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,257 @@
+# rubocop:disable Style/CaseEquality, Style/For, Lint/RedundantCopDisableDirective, Lint/MissingCopEnableDirective, Metrics/ModuleLength, Metrics/MethodLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    arr = self
```
Comment (Forison) [inline enumerable.rb:5]: * [x] There should be a working my_each identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
compare the results of the examples below and make adjustments to your method.
``` my_array = [7, 2, 3, 4, 5, 6, 7, 8] ```
``` p my_array.my_each{ | a | a} != p my_array.each{ | a | a}

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    if block_given?
+      for i in arr do
+        yield i
+      end
+    else
+      arr.to_enum
```
Comment (Forison) [inline enumerable.rb]: * [x] Pass an argument to ```to_enum```, ie to_enum(:name_of_your_method), in this case ```to_enum(:my_each)```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  def my_each_with_index
+    arr = self
+    if block_given?
+      j = 0
+      for i in arr do
+        yield i, j
+        j += 1
+      end
+    else
+      arr.to_enum
```
Comment (Forison) [inline enumerable.rb]: * [x] Pass an argument to ```to_enum```, ie to_enum(:name_of_your_method), in this case ```to_enum(:my_each_with_index)```
after that compare answers.
```p my_array.my_each_with_index ``` and
```p my_array.each_with_index ```

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+  def my_select
+    if block_given?
+      new_arr = []
+      arr = self
+      arr.my_each do |i|
+        new_arr.push(i) if yield i
+      end
+      new_arr
+    else
+      arr.to_enum
```
Comment (Forison) [inline enumerable.rb]: * [x] Pass an argument to to_enum, ie to_enum(:name_of_your_method), in this case to_enum(:my_select)
after that compare answers.
```p my_array.my_select```  and ```p my_array.select```

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    end
+  end
+
+  def my_all?(all_arg = nil)
+    arr = self
```
Comment (Forison) [inline enumerable.rb:42]: * [x] There should be a working `my_all?` dentical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
check the example below, they do not return same answers
```p [1, 2i, 3.14].my_all?(Numeric)```  != ```p [1, 2i, 3.14].all?(Numeric)```

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    end
+    result
+  end
+end
+
+def multiply_els(arr_multiply)
+  arr_multiply.my_inject { |multi, num| multi * num }
+end
+
+# my_array = [7, 2, 3, 4, 5, 6, 7, 8]
```
Comment (Forison) [inline enumerable.rb]: * [x] kindly put your test examples in a separate file inside the root directory.

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Require changes🛑
# 🙋‍♂️ @moinkhanif,
good job on implementing the changes suggested in the previous review. you are almost close to having this project approved, There are still a few more changes to do.
## Check comments below👇. 
# Kindly make the changes and submit the project for review.
### Happy coding
### Addo Forison

**Review by MauricioRobayo — APPROVED** [review]
This is it, you got it!

Congratulations, please move on.

Status: __Approved__.

---

Optional:

* In the [Odin project #12](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2) you were asked to modify your `my_map` method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature.

```ruby
>> my_proc = Proc.new { |num| num > 5 }
>> [4, 6].my_map(my_proc) # should return false true
>> [4, 6].my_map(my_proc) { |num| num < 5 } # should return false true
```

* You have a `rubocop_todo.yml` on your repo 🤔. Files not related to the project should not be committed and added to .gitignore (e.g. `.DS_Store`) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)).

**Pull request merged**